### PR TITLE
Runhouse ssh into local clusters

### PR DESCRIPTION
**Pull request by carolineechen**

### CODE REVIEW

**Comment by carolineechen** [comment]
* **#1012** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1012?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @carolineechen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/resources/hardware/utils.py [hunk 1]
```diff
@@ -129,3 +131,23 @@ def detect_cuda_version_or_cpu(cluster: "Cluster" = None):
...
+        ip=address,
+        ssh_user=ssh_user,
+        port=ssh_port,
+        ssh_private_key=ssh_private_key,
+        docker_user=docker_user,
+    )
+    ssh_command = runner._ssh_base_command(
+        ssh_mode=SshMode.INTERACTIVE, port_forward=None
+    )
+    subprocess.run(ssh_command)
```
Comment (jlewitt1) [inline runhouse/resources/hardware/utils.py:153]: missing `subprocess ` import?

File: runhouse/main.py [hunk 2]
```diff
@@ -115,31 +115,52 @@ def notebook(
...
-    c.ssh()
+
+            return
+
+        resource_handle = state[0].get("handle", {})
+        _run_ssh_command(
+            address=resource_handle.head_ip,
+            ssh_user=resource_handle.ssh_user or "ubuntu",
+            ssh_port=resource_handle.stable_ssh_ports[0] or DEFAULT_SSH_PORT,
+            ssh_private_key=OnDemandCluster.DEFAULT_KEYFILE,
```
Comment (jlewitt1) [inline runhouse/main.py:161]: does this support relative key paths?

File: runhouse/main.py [hunk 3]
```diff
@@ -115,31 +115,52 @@ def notebook(
...
 
-    if not c.is_shared:
-        if up:
-            try:
-                c.up_if_not()
-            except NotImplementedError:
+        if not c.is_shared:
+            if up:
+                try:
+                    c.up_if_not()
```
Comment (jlewitt1) [inline runhouse/main.py:124]:
```suggestion
                    c = c.up_if_not()
```
might be an edge case, but if the cluster isn't actually up and we do bring it back, we don't want to be storing stale IPs, which returning the updated cluster object prevents

File: runhouse/main.py [hunk 4]
```diff
@@ -115,31 +115,52 @@ def notebook(
...
+
+        from runhouse.constants import DEFAULT_SSH_PORT
+        from runhouse.resources import OnDemandCluster
+
+        from runhouse.resources.hardware.utils import _run_ssh_command
+
+        state = sky.status(cluster_names=[cluster_name], refresh=False)
+
+        if len(state) == 0:
+            logger.error(
```
Comment (jlewitt1) [inline runhouse/main.py:149]: `console.print` for consistency?

File: runhouse/main.py [hunk 5]
```diff
@@ -115,31 +115,52 @@ def notebook(
...
-        elif not c.is_up():
-            console.print(
-                f"Cluster {cluster_name} is not up. Please run `runhouse ssh {cluster_name} --up`."
+
+        c.ssh()
+    except ValueError:
+        import sky
+
+        from runhouse.constants import DEFAULT_SSH_PORT
+        from runhouse.resources import OnDemandCluster
```
Comment (jlewitt1) [inline runhouse/main.py:142]:
```suggestion
        from runhouse import OnDemandCluster
```

File: runhouse/main.py (hunk 3 above)
Comment (carolineechen) [reply runhouse/main.py:124]: it looks like `up_if_not` and `up` does directly update the fields of the original cluster object, even if we don't reassign it with the new returned object

File: runhouse/main.py (hunk 2 above)
Comment (carolineechen) [reply runhouse/main.py:161]: both relative and absolute seem to work

**Pull request merged**